### PR TITLE
fix(cli/agent): support Vercel UI Message Stream v1 SSE format

### PR DESCRIPTION
**Pull request by NeOMakinG**
## Summary
- Adds backward-compatible support for Vercel AI SDK UI Message Stream v1 format in `clients/cli/src/agent/client.ts`
- Required to keep the agent-backend integration test passing after [agent-backend#119](https://github.com/vultisig/agent-backend/pull/119) lands

## Why
The agent-backend is migrating to the Vercel AI SDK v1 stream format which sends bare `data:` lines (SSE-spec default event = "message") with the type encoded in the JSON `type` field, instead of `event: <name>\ndata: <json>` pairs. The existing CLI parser silently ignored these, causing the agent backend to "appear connected but not generate responses" in the integration test.

## How
When `event === "message"` (the SSE default) and the payload has a `type` field, map v1 types to legacy event names and unwrap nested `data` payloads:
- `text-delta` → `text_delta`
- `data-title` → `title` (with `data` unwrapped)
- `data-message` → `message`
- `data-actions` → `actions`
- `finish` → `done`
- `error` → `error` (with `errorText` → `error`)

Existing `event: text_delta\ndata: ...` legacy format still works unchanged.

## Test plan
- [ ] CI passes
- [ ] Integration test on agent-backend#119 passes after this is merged

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Bug Fixes

- Improved Server-Sent Event parsing to support newer message stream protocol versions
- Enhanced event mapping and payload handling for better data compatibility
- Better error message normalization across different protocol formats
- Added verbose mode diagnostics for event processing troubleshooting

### CODE REVIEW

File: clients/cli/src/agent/client.ts [hunk 1]
```diff
@@ -240,9 +240,35 @@ export class AgentClient {
...
-      const parsed = JSON.parse(data)
+      const rawParsed = JSON.parse(data)
+
+      // Vercel UI Message Stream v1 adapter: when no `event:` line preceded
+      // the data line, the SSE spec defaults `event` to "message". The new
+      // backend uses bare `data:` lines with the type encoded in the JSON
+      // `type` field. Map v1 types to legacy event names + payload shape so
+      // the existing switch keeps working.
+      let resolvedEvent = event
+      let parsed = rawParsed
```
Comment (gomesalexandre) [inline clients/cli/src/agent/client.ts:251]: q: the v1 gate is "SSE default event + top-level `type:string`". works today because `ConversationMessage` has no top-level `type` field (see `types.ts:98`). worth a one-line comment pinning that the gate depends on that absence, so the next person adding a `type` field to `ConversationMessage` has a chance of noticing they just split the parser in two.

not blocking. structurally fine for a coupled cli/backend.

File: clients/cli/src/agent/client.ts [hunk 2]
```diff
@@ -240,9 +240,35 @@ export class AgentClient {
...
+      if (event === 'message' && typeof rawParsed?.type === 'string') {
+        const v1Type = rawParsed.type as string
+        if (v1Type === 'text-delta') {
+          resolvedEvent = 'text_delta'
+        } else if (v1Type === 'finish') {
```
Comment (gomesalexandre) [inline clients/cli/src/agent/client.ts:261]: q: Vercel v1 `finish` carries `finishReason` and `usage`. collapsing to `done` drops both. fine for today since `UsageInfo` exists in `types.ts` but isn't surfaced by the cli. flagging for the day chat ux wants token counts, so we don't re-parse the stream later.

File: clients/cli/src/agent/client.ts [hunk 3]
```diff
@@ -240,9 +240,35 @@ export class AgentClient {
...
+          resolvedEvent = 'done'
+        } else if (v1Type === 'error') {
+          resolvedEvent = 'error'
+          parsed = { error: rawParsed.errorText ?? rawParsed.error }
+        } else if (v1Type.startsWith('data-')) {
```
Comment (gomesalexandre) [inline clients/cli/src/agent/client.ts:270]: preferably-blocking (at the fixture level, not this pr): confirm agent-backend#119 emits `data-tx_ready` (snake) and not `data-tx-ready` (kebab, which is the natural Vercel convention for data-* parts). `slice(5)` preserves whatever follows `data-`, so a kebab producer silently routes to no matching switch case and the tx drops on the floor.

cheapest fix: pin a fixture in agent-backend#119's integration test that covers at least `data-tx_ready`, `data-actions`, `data-title`, `data-message`. if the wire format drifts later, ci screams.

File: clients/cli/src/agent/client.ts [hunk 4]
```diff
@@ -240,9 +240,35 @@ export class AgentClient {
...
+          resolvedEvent = v1Type.slice(5) // data-title → title, etc.
+          if (rawParsed.data && typeof rawParsed.data === 'object') {
```
Comment (gomesalexandre) [inline clients/cli/src/agent/client.ts]: suggestion: `typeof rawParsed.data === 'object'` matches arrays too. if backend ever sends `{type: 'data-suggestions', data: [...]}` (naked array, no wrapper), unwrap produces `parsed = [...]`, then `case 'suggestions'` does `parsed.suggestions || []` which is `undefined || []` and suggestions silently drop.

either:
- guard with `!Array.isArray(rawParsed.data)`, OR
- document in agent-backend#119 that `data-*` payloads are always object-shaped wrappers.

same outcome either way. pick whichever boundary you want to own.

File: clients/cli/src/agent/client.ts [hunk 5]
```diff
@@ -240,9 +240,35 @@ export class AgentClient {
...
+            parsed = rawParsed.data
+          }
+        } else {
```
Comment (gomesalexandre) [inline clients/cli/src/agent/client.ts:283]: preferably-blocking: this is the one real durable finding from the 5 passes. Vercel v1 has native parts the backend may emit without announcing (`tool-input-available`, `tool-output-available`, `tool-output-error`, `reasoning`, `start`, `start-step`, `finish-step`, `text-start`, `text-end`). the adapter maps none of them. they all resolve to a string that doesn't match any switch case and drop on the floor. `onToolProgress` goes dark the moment backend migrates tool progress to v1 native parts.

cheapest fix right here:

```ts
} else {
  resolvedEvent = v1Type
  if (this.verbose) process.stderr.write(`[SSE:unmapped v1 type: ${v1Type}]\n`)
}
```

same bar as the existing verbose logs for `tool_progress` / `actions` / `tx_ready`. makes backend drift visible instead of invisible. you'll see it in `--verbose` the first time it happens, no guessing.

not asking to map `tool-input-available` etc preemptively. that's yagni until backend commits. just the log.

**Review by gomesalexandre — COMMENTED** [review]
## conceptual review only (functional/runtime cli test coming separately)

5 passes: Claude deep read, Codex adversarial (block), Claude contradversial (steel-man the ship-it case), Codex arbitration (ship-with-comments), Claude final synth.

tl;dr this is a migration shim, not a protocol framework project. the ship-it side wins on architecture and scope. the adversarial side overreaches on "extract a versioned normalizer + protocol negotiation handshake" which is cosmic overkill for a shim between two co-owned services whose contract IS the agent-backend#119 integration test.

**one durable concern survives both pressures: silent failure modes.**

the catch-all `resolvedEvent = v1Type`, the array-shaped `data` unwrap, the snake-vs-kebab ambiguity on `data-*`, and the finish metadata loss are all "silent today, surprising later" risks. not one is block-level on its own. collectively they mean the cli can look connected and emit nothing when the backend drifts.

**proportional fix:** verbose-log unmapped v1 types at the same bar as the existing `[SSE:tool_progress]` logs already in this file. same cost, fails loud.

line-level notes below. not blocking.

risk: low, migration shim, legacy branch unchanged. ship when ci is green + agent-backend#119 pins the wire format with a fixture per `data-*` branch (snake, not kebab).

File: clients/cli/src/agent/client.ts (hunk 1 above)
Comment (NeOMakinG) [reply clients/cli/src/agent/client.ts:251]: Done in f8febe91 — added a 5-line comment at the v1-gate noting that this relies on `ConversationMessage` (types.ts:98) having no top-level `type` field, and that adding one would require re-splitting the parser.

File: clients/cli/src/agent/client.ts (hunk 2 above)
Comment (NeOMakinG) [reply clients/cli/src/agent/client.ts:261]: Done in 3aa3321d — added a comment above the `finish` → `done` mapping flagging that `finishReason` and `usage` are dropped, so we don't re-parse the stream when the CLI eventually surfaces token usage.

File: clients/cli/src/agent/client.ts (hunk 3 above)
Comment (NeOMakinG) [reply clients/cli/src/agent/client.ts:270]: Done in 44ae72b8 — inline comment on the `slice(5)` line noting that we expect snake_case after `data-` and that the contract is pinned by the agent-backend#119 fixture. Agree the real guarantee belongs in that fixture, not here — @gomesalexandre ok to track the fixture-level assertion in agent-backend#119?

File: clients/cli/src/agent/client.ts (hunk 4 above)
Comment (NeOMakinG) [reply clients/cli/src/agent/client.ts]: Fixed in 010f49df — added `!Array.isArray(rawParsed.data)` to the unwrap guard so a naked-array payload no longer silently collapses to `parsed = [...]`. If the backend ever sends that shape, it now falls through to the switch with `parsed` still pointing at the v1 envelope and the verbose unmapped-type log from #5 will surface it.

File: clients/cli/src/agent/client.ts (hunk 5 above)
Comment (NeOMakinG) [reply clients/cli/src/agent/client.ts:283]: Fixed in 78d27365 — added exactly the suggested verbose log:

```ts
} else {
  resolvedEvent = v1Type
  if (this.verbose) process.stderr.write(`[SSE:unmapped v1 type: ${v1Type}]\n`)
}
```

Style matches the existing `[SSE:tool_progress]` / `[SSE:actions]` / `[SSE:tx_ready]` logs in `handleSSEEvent`. No preemptive mapping of `tool-input-available` / `reasoning` / `start-step` — YAGNI until backend commits, agreed.

**Comment by NeOMakinG** [comment]
Addressed all 5 review comments — branch pushed.

- L251 — pinned v1-gate assumption comment — f8febe91
- L256 — documented finish-metadata loss — 3aa3321d
- L261 — snake_case expectation comment — 44ae72b8
- L263 — `!Array.isArray` guard on data unwrap — 010f49df
- L266 — verbose log for unmapped v1 types — 78d27365

Re L261: the real fixture pin belongs in agent-backend#119 — @gomesalexandre ok to track there?

### Verification
- `yarn workspace @vultisig/cli typecheck` — no errors on `client.ts` (the 4 unrelated errors on `executor.ts:495/502/972/978` are pre-existing on `main`)
- `yarn workspace @vultisig/cli test` — 47/47 passing (including the 12 `client.test.ts` tests)
- `yarn eslint clients/cli/src/agent/client.ts` — clean
- `yarn prettier --check clients/cli/src/agent/client.ts` — clean

— vultisig-ops

**Review by gomesalexandre — APPROVED** [review]
## round 2 re-review - pulled 78d27365, ran cli tests locally

retest with cli per ask: fresh worktree on `78d27365`, `yarn install`, `yarn test:cli` green (47 pre-existing tests still passing, no regressions to the legacy path), then authored **14 v1-adapter-specific tests** to empirically validate each branch of the new code.

**all 14 v1 tests pass**, including the edge cases from round 1:

- v1 `text-delta`, `data-title`, `data-actions`, `data-message`, `data-tx_ready`: mapping + unwrap work
- v1 `error` both paths (`errorText` + legacy `error` fallback via `??`): work
- v1 `finish` -> `done`: collapses cleanly, no throw
- v1 unmapped types (`text-start`, `reasoning`): verbose log fires with `[SSE:unmapped v1 type: ...]` as specified ✓
- v1 naked-array `data-suggestions` (the shape the new guard protects): guard prevents the `parsed = [...]` assignment, but result is still `[]` silently (nuance below)
- legacy `event: message` path unchanged: no regression
- kebab `data-tx-ready`: silently drops (now pinned as a test, pending the fixture in agent-backend#119)
- mixed legacy + v1 streams: interleave correctly

**one nuance worth flagging** (under the array-guard thread): the guard does prevent the pathological `parsed = [...]` assignment, but the switch case reading `parsed.suggestions` still silently yields `[]` because `parsed` becomes the v1 envelope instead of the expected wrapper. net effect is the same drop, just via a different path. the guard still wins on debuggability (envelope stays inspectable under `--verbose` with a breakpoint) so keeping it green. not a blocker, empirically confirmed.

**approving.** shim is the right shape, fixes are proportional, empirically covered.

happy to PR the 14 v1 tests as a follow-up if useful. kept them local to avoid pushing to your branch uninvited.

risk: low. regression surface tested. legacy path intact.

File: clients/cli/src/agent/client.ts (hunk 1 above)
Comment (gomesalexandre) [reply clients/cli/src/agent/client.ts:251]: perfect, comment lands exactly where future-me needs it. ty for the `types.ts:98` back-ref. closing.

File: clients/cli/src/agent/client.ts (hunk 2 above)
Comment (gomesalexandre) [reply clients/cli/src/agent/client.ts:261]: ty, yagni-ed correctly. verified locally: `{"type":"finish","finishReason":"stop","usage":{...}}` collapses to done cleanly, no throw. revisit when ui wants tokens.

File: clients/cli/src/agent/client.ts (hunk 3 above)
Comment (gomesalexandre) [reply clients/cli/src/agent/client.ts:270]: yes, tracking in agent-backend#119 is the right home. cli can't own the wire format unilaterally.

pinned locally with a test case for both snake (`data-tx_ready` routes to `tx_ready`, tx accumulates) and kebab (`data-tx-ready` silently drops, current behavior). makes the assumption a ci-visible contract, but agent-backend#119 should still own the producer side.

closing.

File: clients/cli/src/agent/client.ts (hunk 4 above)
Comment (gomesalexandre) [reply clients/cli/src/agent/client.ts]: one nuance to correct before this fossilizes, not a blocker:

> it now falls through to the switch with `parsed` still pointing at the v1 envelope and the verbose unmapped-type log from #5 will surface it

the verbose unmapped-type log won't fire here. verified empirically with a test: a naked-array `{type:"data-suggestions", data:[...]}` still takes the `startsWith('data-')` branch, still sets `resolvedEvent = 'suggestions'`, still hits `case 'suggestions'` in the switch. the unmapped-type log only fires in the `else` branch (unknown top-level `type`).

so:
- before guard: `parsed = [...]`, `parsed.suggestions` is `undefined`, silent empty
- after guard: `parsed = {type, data:[...]}`, `parsed.suggestions` is `undefined`, silent empty

net effect is the same drop, just via a different path. the guard does win on one thing: `parsed` stays inspectable as the envelope under a `--verbose` breakpoint if you're debugging, instead of an opaque array. worth having.

keeping it green because the shape is purely hypothetical and backend has no reason to emit naked arrays. but if it ever does, neither the guard nor the log catches it, only the switch case's own defensive coding would.

ok to close. just wanted the next reader not to think the log saves us here.

File: clients/cli/src/agent/client.ts (hunk 5 above)
Comment (gomesalexandre) [reply clients/cli/src/agent/client.ts:283]: nice, exact snippet. verified locally: `text-start` and `reasoning` frames produce `[SSE:unmapped v1 type: text-start]` / `[SSE:unmapped v1 type: reasoning]` under `--verbose`. ship it.

**Review by rcoderdev — APPROVED** [review]

File: clients/cli/src/agent/client.ts (hunk 4 above)
Comment (NeOMakinG) [reply clients/cli/src/agent/client.ts]: right, you're correct — the log gates on the `else` branch (unknown top-level type), not the switch fallthrough. so a naked-array `data-suggestions` still silently drops; the guard just makes `parsed` inspectable as the envelope under `--verbose`, which is worth having but isn't the "fails loud" win i claimed. will note it for the next reader. thanks for the catch, closing.
